### PR TITLE
feat:support cypher parameter

### PR DESCRIPTION
**Pull request by hetao92**

### CODE REVIEW

File: service/dao/dao.go [hunk 1]
```diff
@@ -287,26 +287,29 @@ func Disconnect(nsid string) {
 	pool.Disconnect(nsid)
 }
 
-func Execute(nsid string, gql string) (result ExecuteResult, err error) {
+func Execute(nsid string, gql string, paramList common.ParameterList) (result ExecuteResult, cmdResult string, err error) {
```
Comment (nianiaJR) [inline service/dao/dao.go:290]: paramList can be null? so use `paramList ...`?

Comment (nianiaJR) [inline service/dao/dao.go:290]: cmdResult? what's it like?

File: service/pool/pool.go [hunk 2]
```diff
@@ -126,13 +277,40 @@ func NewConnection(address string, port int, username string, password string) (
...
-							err = ConnectionClosedError
+						cmdResult := ""
+						if len(request.ParamList) > 0 {
+							err = executeCmd(request.ParamList, &connection.parameterMap)
+							if err != nil {
+								cmdResult = err.Error()
+							} else {
+								cmdResult = SetParamsSuccess
+							}
 						}
```
Comment (nianiaJR) [inline service/pool/pool.go:288]: It's like that the `cmdResult` is not necessary

Comment (nianiaJR) [reply service/pool/pool.go:288]: If it's error, just return error. The channel response result just return `nil, error` enough

**Review by nianiaJR — APPROVED** [review]
LGTM

File: service/dao/dao.go [hunk 3]
```diff
@@ -287,26 +287,30 @@ func Disconnect(nsid string) {
 	pool.Disconnect(nsid)
 }
 
-func Execute(nsid string, gql string) (result ExecuteResult, err error) {
+func Execute(nsid string, gql string, paramList common.ParameterList) (result ExecuteResult, params common.ParameterMap, err error) {
```
Comment (veezhang) [inline service/dao/dao.go:290]: Why not put `params` into `ExecuteResult`?

File: controllers/db.go [hunk 4]
```diff
@@ -84,14 +86,19 @@ func (this *DatabaseController) Execute() {
...
 			res.Data = &result
 		} else {
 			res.Code = -1
 			res.Message = err.Error()
 		}
+		if len(paramsMap) == 0 {
+			res.Params = nil
+		} else {
+			res.Params = paramsMap
+		}
```
Comment (veezhang) [inline controllers/db.go:101]: The default value of `res.Params` is `nil`, need not to assign with `nil`.
```go
if len(paramsMap) > 0 {
    res.Params = paramsMap
}
```

File: service/pool/pool.go [hunk 5]
```diff
@@ -12,11 +15,20 @@ import (
 
 	uuid "github.com/satori/go.uuid"
 	nebula "github.com/vesoft-inc/nebula-go/v2"
+	nebulaType "github.com/vesoft-inc/nebula-go/v2/nebula"
```
Comment (veezhang) [inline service/pool/pool.go:18]: The pkg name alias seems that they are all lowercase, and I haven't seen any uppercase ones.

File: service/pool/pool.go [hunk 6]
```diff
@@ -76,6 +91,203 @@ func isThriftTransportError(err error) bool {
...
+	return showMap, nil
+}
+
+func defineParams(args string, parameterMap *common.ParameterMap) (err error) {
+	argsRewritten := strings.Replace(args, "'", "\"", -1)
+	reg := regexp.MustCompile(`^\s*:param\s+(\S+)\s*=>(.*)$`)
+	if reg == nil {
+		err = errors.New("invalid regular expression")
+		return
+	}
```
Comment (veezhang) [inline service/pool/pool.go:227]: The `reg` cannot be `nil`.

File: service/pool/pool.go [hunk 7]
```diff
@@ -76,6 +91,203 @@ func isThriftTransportError(err error) bool {
...
+	res := matchResult[0]
+	/*
+	 * :params -> [":params",""]
+	 * :params p1 -> ["params","p1"]
+	 */
+	if len(res) != 2 {
+		return
+	} else {
```
Comment (veezhang) [inline service/pool/pool.go:274]: useless `else`.

File: service/pool/pool.go [hunk 8]
```diff
@@ -76,6 +91,203 @@ func isThriftTransportError(err error) bool {
...
+	}
+	isLocal = true
+	words := strings.Fields(plain[1:])
+	localCmdName := words[0]
+	switch strings.ToLower(localCmdName) {
+	case "param":
+		{
+			localCmd = Param
+			args = []string{plain}
+		}
```
Comment (veezhang) [inline service/pool/pool.go:186]: useless `{` and `}`.

**Review by veezhang — APPROVED** [review]
LGTM

**Pull request merged**